### ISSUE TITLE
Text measing API?

### ISSUE THREAD

**Issue by wm4** [libass/libass#348]
Are there any thought about text measuring APIs? I'd like to have something that can retrieve the typographic measurements of some text. Any ideas whether this would be an acceptable thing, and how to expose it as API? What measurements should it even return? (Typography is hard.)

**Comment by CoffeeFlux** [comment]
Proper text measuring APIs would definitely be useful. Currently SubInspector does [this](https://github.com/TypesettingTools/SubInspector/blob/master/src/SubInspector.c#L268) for bounds detection, but a more proper method would be welcome. Related: https://github.com/libass/libass/pull/329

**Comment by wm4** [comment]
Yes, computing the bounding box is possible, but kind of sucks. It's slower than necessary, and not typographically correct.
I also assume it could skip full rendering of the measured text. It will often happen that the library user measures text, and then renders only a part of it, or so.

**Comment by wm4** [comment]
I hoped for some discussion about possible APIs. I guess I'll just make a PR one day with an API that suits my needs.

**Comment by MrSmile** [comment]
There are two different types of text measurement: logical (text advance + font height and depth) and graphical (bbox of actual vector outline). The first is useful for alignment, tables and such; the second for allocating bitmaps. The first is also much cheaper to compute (just dump Harfbuzz output) but second makes sense in case of transformed text.

So, what kind of measurement are you thinking of?

**Comment by astiob** [comment]
Ideally we would be able to output both. Also, intercharacter spacing (both from the font and from `\fsp`) may be included or excluded on the last (right in left-to-right) edge.

**Comment by wm4** [comment]
Thanks for giving a thought about it. I'm actually not sure myself what makes sense.

I want measurements for layouting and input feedback purposes. For example, I'd want to compute the "logical" bounding box of text to draw a frame around it. A bitmap bounding box could return a tiny bounding box for "." (instead of the logical line height), or an arbitrary box (higher than the text height) if the text exceeds the line bounds somehow (think of "zalgo"). And of course it doesn't tell you the baseline. At least at this point, using the ASS property that the font size equals the "line height" stops being sufficient for vertical alignment. Or maybe I'm overthinking, and returning baseline and line count is all what is missing?

Moreover, I don't just want bounding boxes of the entire text. I also want to:
- map coordinates back to a character/codepoint position (think of mouse selection of text)
- return grapheme cluster boundaries and order
- know the bounds of a certain character or text run (this could result in multiple rectangles if text is wrapped)
- have enough information to perform ellipsization
- know how many lines some text is (accounting for all weird methods you could do line breaks in ASS, but also when screen bounds or margins cause the text to wrap)

Some of these can probably be handled by the same mechanisms. Also, to return less data from the API and to make mapping between bytes and logical entities, I thought maybe a new special tag could be introduced, that "probes" boundaries of enclosed text, or if the text is empty, the position a cursor would take:

```{\start_bounds(123)}some text{\end_bounds}```

Where start_bounds takes an arbitrary user-set ID as argument. Each bounds run would return one or multiple rectangles. The C API might look something like this:

```
// Measures the text that an equivalent ass_render_frame would draw.
// res is an array of ASS_Bounds structs, and res_size is the maximum number of the array
// items that libass can write to. The function returns the number of items that were written to
// (or a negative value on error). If the return value is the same as res_size, there might have
// been not enough ASS_Bounds item to report all results.
int ass_measure(ASS_Renderer *priv, ASS_Track *track, long long now, ASS_Bounds *res, int res_size);

struct ASS_Bounds {
  // ID from the start_bounds tag. If the text got broken into multiple parts (like 
  // due to text wrapping), there may be multiple ASS_Bounds item with the same
  // ID.
  int id;
  // Line number, after text wrapping. 0 is the first line.
  int line;
  // Logical bounding box.
  float l, t, r, b; 
  // Text baseline relative to t.
  float baseline;
};
```

The user could easily determine the physical location of text fragments using this API. To determine the line count, you could use a start_bounds tag with an empty "probe" text range at the very end of the event. To determine the total bounds, the entire event text could be put into a single start_bounds text (which would also give you per-line information). We wouldn't allow nesting of bounds tags, so unfortunately you'd need to call this function multiple times if you want to know both the bounds of the entire text, and the bounds of some fragment contained within it.

Instead of this starts_bounds tag, this could be a more down-to-earth API that somehow makes every grapheme cluster or so into its own ASS_Bounds, and has a byte offset/count field instead of an id field. But then we'd probably need to return per-codepoint info, which assigns each codepoint to a grapheme cluster ID (what harfbuzz seems to do?). The glyph position would be given by ASS_Track index + byte offset into the ASS_Track text. Each glyph would have a cluster index, which index into a separate cluster array, which has the bounding box of each cluster.

Ellipsization could probably be a separate feature that is done by the renderer, maybe a new wrap mode.

Most of these things don't really have any use for a pure subtitle renderer. Is there anything you'd clearly consider out of scope and would not want to have in libass?

Also possible that this is too complicated as that I could implement at all.

**Comment by wm4** [comment]
Also pinging @grigorig, maybe he's still alive and he has experience with harfbuzz.

**Comment by CounterPillow** [comment]
With harfbuzz, one can get the width and height of a certain string of text after shaping, provided one first splits up the string whenever it switches *font style* or noiɈɔɘɿib. For more about this, the section [what harfbuzz doesn't do](https://harfbuzz.github.io/what-harfbuzz-doesnt-do.html) can be helpful.

**For the width:**

After shaping, call [hb_buffer_get_glyph_positions](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-get-glyph-positions) on the buffer. This will give you an array of [hb_glyph_position_t](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-glyph-position-t-struct) which contain information such as the x offset, the y offset, and how far this advances the offset in x and y directions. To get the total width for our text, we need to find the min and the max of the x offset, and then add x_advance to the max as we want the coordinate _after_ the last glyph to get a bounding box. Depending on the used writing system's direction, this is probably the first or last glyph in the array.

**For the height:**

Same concept, but I presume here we need to actually figure out the min and max from each glyph, unless the text directionality is vertical in which case the method for width and height are flipped.
